### ISSUE TITLE
使用unsloth微调后的模型能否转换为rkllm模型

### ISSUE THREAD

**Issue by hsj1104** [airockchip/rknn-llm#386]
`python3 export_rkllm.py
/home/chicken/miniconda3/envs/rkllm/lib/python3.11/site-packages/rkllm/api/rkllm.py:5: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  self.base = RKLLMBase(v)
INFO: rkllm-toolkit version: 1.2.1
Building model: 100%|??????????????????????| 2331/2331 [00:04<00:00, 502.95it/s]
Downloading data files: 100%|??????????????????| 1/1 [00:00<00:00, 30615.36it/s]
Extracting data files: 100%|????????????????????| 1/1 [00:00<00:00, 4013.69it/s]
Generating train split: 21 examples [00:00, 10484.51 examples/s]
Optimizing model:   0%|                                  | 0/28 [00:00<?, ?it/s]
ERROR: Catch exception when optimizing model: 'mlp.down_proj'
Build model failed!`
我用原来没有微调的模型是可以正常转换的，但是使用unsloth微调之后，构建模型就会报上面的错误

**Comment by waydong** [comment]
你好，微调的哪个模型？

**Comment by hsj1104** [comment]
DeepSeek-R1-Distill-Qwen-1.5B，数据集用的是开源的medical-o1-reasoning-SFT，工具链 是 1.2.1版本 ，python版本  3.11

**Comment by waydong** [comment]
模型结构的函数名有更改吗，可以比对下微调前后的模型文件。

**Comment by hsj1104** [comment]
没有改过模型结构，只是 用新的数据集又重新训练 了一下模型，这是代码：
`from unsloth import FastLanguageModel
from modelscope.msdatasets import MsDataset
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

train_prompt_style = """Below is an instruction that describes a task, paired with an input that provides further context. 
Write a response that appropriately completes the request. 
Before answering, think carefully about the question and create a step-by-step chain of thoughts to ensure a logical and accurate response.

### Instruction:
You are a medical expert with advanced knowledge in clinical reasoning, diagnostics, and treatment planning. 
Please answer the following medical question. 

### Question:
{}

### Response:
<think>
{}
</think>
{}"""

max_seq_length = 2048 
dtype = None 
load_in_4bit = False

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "./DeepSeek-R1-Distill-Qwen-1.5B",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
)

EOS_TOKEN = tokenizer.eos_token  # Must add EOS_TOKEN

def formatting_prompts_func(examples):
    inputs = examples["Question"]
    cots = examples["Complex_CoT"]
    outputs = examples["Response"]
    texts = []
    for input, cot, output in zip(inputs, cots, outputs):
        text = train_prompt_style.format(input, cot, output) + EOS_TOKEN
        texts.append(text)
    return {
        "text": texts,
    }

ds =  MsDataset.load('AI-ModelScope/medical-o1-reasoning-SFT', split = "train")
dataset = ds.map(formatting_prompts_func, batched = True,)
dataset["text"][0]

model = FastLanguageModel.get_peft_model(
    model,
    r=16,  
    target_modules=[
        "q_proj",
        "k_proj",
        "v_proj",
        "o_proj",
        "gate_proj",
        "up_proj",
        "down_proj",
    ],
    lora_alpha=16,
    lora_dropout=0,  
    bias="none",  
    use_gradient_checkpointing="unsloth",  # True or "unsloth" for very long context
    random_state=3407,
    use_rslora=False,  
    loftq_config=None,
)

trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=dataset,
    dataset_text_field="text",
    max_seq_length=max_seq_length,
    dataset_num_proc=2,
    args=TrainingArguments(
        per_device_train_batch_size=2,
        gradient_accumulation_steps=4,
        num_train_epochs = 3,
        warmup_steps=5,
        # max_steps=60,
        learning_rate=2e-5,
        fp16=not is_bfloat16_supported(),
        bf16=is_bfloat16_supported(),
        logging_steps=10,
        optim="adamw_torch_fused",
        weight_decay=0.01,
        lr_scheduler_type="linear",
        seed=3407,
        output_dir="outputs",
    ),
)

#wandb.init()
trainer_stats = trainer.train()

new_model_local = "DeepSeek-R1-Medical"
model.save_pretrained(new_model_local) 
tokenizer.save_pretrained(new_model_local)
model.save_pretrained_merged(new_model_local, tokenizer, save_method = "merged_16bit",)
`

**Comment by waydong** [comment]
1、要看转出来的模型文件，2、也可以试试1.2.2版本

**Comment by hsj1104** [comment]
具体要看哪个部分？权重文件是safetensors格式的

**Comment by waydong** [comment]
你先用1.2.2版本转换看看。

**Comment by hsj1104** [comment]
1.2.2版本的examples我没看到有DeepSeek-R1-Distill-Qwen-1.5B的，直接用1.2.1版本的代码转换 ？

**Comment by hsj1104** [comment]
`$ python3 export_rkllm.py
/home/chicken/miniconda3/envs/rkllm/lib/python3.11/site-packages/rkllm/api/rkllm.py:5: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  self.base = RKLLMBase(v)
INFO: rkllm-toolkit version: 1.2.2
Building model: 100%|??????????????????????| 2331/2331 [00:05<00:00, 461.96it/s]
Generating train split: 21 examples [00:00, 3466.64 examples/s]
Optimizing model:   0%|                                  | 0/28 [00:00<?, ?it/s]
ERROR: Catch exception when optimizing model: 'mlp.down_proj'
Build model failed!
`
1.2.2版本还是报相同的错误

**Comment by hsj1104** [comment]
1.export HF_ENDPOINT=https://hf-mirror.com
2.huggingface-cli download --resume-download deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B --local-dir ./DeepSeek-R1-Distill-Qwen-1.5B/
3.python test.py
``from unsloth import FastLanguageModel
from modelscope.msdatasets import MsDataset
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

train_prompt_style = """Below is an instruction that describes a task, paired with an input that provides further context.
Write a response that appropriately completes the request.
Before answering, think carefully about the question and create a step-by-step chain of thoughts to ensure a logical and accurate response.

Instruction:
You are a medical expert with advanced knowledge in clinical reasoning, diagnostics, and treatment planning.
Please answer the following medical question.

Question:
{}

Response:
{} {}"""
max_seq_length = 2048
dtype = None
load_in_4bit = False

model, tokenizer = FastLanguageModel.from_pretrained(
model_name = "./DeepSeek-R1-Distill-Qwen-1.5B",
max_seq_length = max_seq_length,
dtype = dtype,
load_in_4bit = load_in_4bit,
)

EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN

def formatting_prompts_func(examples):
inputs = examples["Question"]
cots = examples["Complex_CoT"]
outputs = examples["Response"]
texts = []
for input, cot, output in zip(inputs, cots, outputs):
text = train_prompt_style.format(input, cot, output) + EOS_TOKEN
texts.append(text)
return {
"text": texts,
}

ds = MsDataset.load('AI-ModelScope/medical-o1-reasoning-SFT', split = "train")
dataset = ds.map(formatting_prompts_func, batched = True,)
dataset["text"][0]

model = FastLanguageModel.get_peft_model(
model,
r=16,
target_modules=[
"q_proj",
"k_proj",
"v_proj",
"o_proj",
"gate_proj",
"up_proj",
"down_proj",
],
lora_alpha=16,
lora_dropout=0,
bias="none",
use_gradient_checkpointing="unsloth", # True or "unsloth" for very long context
random_state=3407,
use_rslora=False,
loftq_config=None,
)

trainer = SFTTrainer(
model=model,
tokenizer=tokenizer,
train_dataset=dataset,
dataset_text_field="text",
max_seq_length=max_seq_length,
dataset_num_proc=2,
args=TrainingArguments(
per_device_train_batch_size=2,
gradient_accumulation_steps=4,
num_train_epochs = 3,
warmup_steps=5,
# max_steps=60,
learning_rate=2e-5,
fp16=not is_bfloat16_supported(),
bf16=is_bfloat16_supported(),
logging_steps=10,
optim="adamw_torch_fused",
weight_decay=0.01,
lr_scheduler_type="linear",
seed=3407,
output_dir="outputs",
),
)

#wandb.init()
trainer_stats = trainer.train()

new_model_local = "DeepSeek-R1-Medical"
model.save_pretrained(new_model_local)
tokenizer.save_pretrained(new_model_local)
model.save_pretrained_merged(new_model_local, tokenizer, save_method = "merged_16bit",)
``

**Comment by waydong** [comment]
可以提供个huggingface模型吗，目前看不出啥问题

**Comment by hsj1104** [comment]
文件太大了 ，只能放到百度网盘 
链接: https://pan.baidu.com/s/1K_FYYZH7hvQJeZeqWPl7kA?pwd=21qe 提取码: 21qe
`ret = llm.load_huggingface(model=modelpath, model_lora = modelpath, device='cuda', dtype="bfloat16", custom_config=None, load_weight=True)`
如果把 model_lora = None  就可以正常构建

**Comment by hsj1104** [comment]
@waydong 有没有找到是什么问题？

**Comment by hsj1104** [comment]
`ret = llm.load_huggingface(model=modelpath, model_lora = lorapath, device='cuda', dtype="bfloat16", custom_config=None, load_weight=True)`
model和model_lora分开两个路径就可以🙂

**Comment by waydong** [comment]
> `ret = llm.load_huggingface(model=modelpath, model_lora = lorapath, device='cuda', dtype="bfloat16", custom_config=None, load_weight=True)` model和model_lora分开两个路径就可以🙂

是的，lora要和base model分开两个文件夹。

**Comment by waydong** [comment]
Closing, feel free to reopen if you'd like to continue.
